### ISSUE TITLE
Multiple targets with automated signed commits for GitHub Actions creates duplicate commits

### ISSUE THREAD

**Issue by olblak** [updatecli/updatecli#2195]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When using the commitapi set to true with a manifest like this one https://github.com/olblak/kubernetes-marketplace/blob/master/.github/updatecli/updatecli.d/kubewarden.yaml

Updatecli creates duplicate commits where each Updatecli run even though all duplicated commits looks valid 
For example this pullrequest https://github.com/olblak/kubernetes-marketplace/pull/118/commits contains the two following commits which are both identical

* https://github.com/olblak/kubernetes-marketplace/pull/118/commits/5378da081cc57adbdea744670563184032d93dc2
* https://github.com/olblak/kubernetes-marketplace/pull/118/commits/32cabe6345b02db2118db35ca2830d8c8cdbe57a

### Expected Behavior

I expect to only see four commit

### Environment

```markdown
- OS:linux
- updatecli: 0.76.0
```

**Comment by adeck-mdb** [comment]
## WARNING! This is an incredibly serious bug.

I ran into this bug in my own workflow, which led me to this issue ticket.
These aren't "duplicate commits". Not really. If you think about how git works, two identical commits is impossible, **unless** something happened between the two commits that essentially reverted the first one.

If you look at each of the commits in [the PR you linked](https://github.com/olblak/kubernetes-marketplace/pull/118), the bug here becomes obvious.
The problem is that you have multiple targets within the same file. So in other words...

1. First, [this commit](https://github.com/olblak/kubernetes-marketplace/pull/118/commits/23bc1bde35bccc139bfcfc58934ca28b5edabf56) updated controller chart version from 2.0.10 -> 2.0.11.
2. Then, [this next commit](https://github.com/olblak/kubernetes-marketplace/pull/118/commits/32cabe6345b02db2118db35ca2830d8c8cdbe57a) updated the defaults chart version from 1.9.4 to 2.0.0. But it didn't just do that. It **also** did something unexpected- it changed controller chart version **back** from 2.0.11 -> 2.0.10. It reverted the previous change.
3. Next, we get [this commit](https://github.com/olblak/kubernetes-marketplace/pull/118/commits/e554d8a2ce77527b1b636dfb7603e0ae84aab6b2) which bumps the CRDs chart version from 1.4.6 -> 1.5.0, **but** rolls back the defaults chart version from 2.0.0 -> 1.9.4.

This is a **very** bad bug. Showstopping. Because chances are pretty good that, if two changes are both in the same file, that means those changes need to be consistent with one another- e.g. client version and server version, or chart version and underlying application version.

**Comment by olblak** [comment]
Thanks for highlighting this issue.
This feature was introduced as experimental in https://github.com/updatecli/updatecli/pull/2070
The tricky part is commits are pushed to GitHub via the api and the local directory get out of sync

**Comment by argoyle** [comment]
Could it be "as simple as" just updating the local repository after the commit has been made using the API? I cloned the repository and started looking into it but it takes a while to understand the flows.

**Comment by olblak** [comment]
> Could it be "as simple as" just updating the local repository after the commit has been made using the API? I cloned the repository and started looking into it but it takes a while to understand the flows.

I thought It was already done before each target we run a git checkout ([link](https://github.com/updatecli/updatecli/blob/d752c2dbb681613b45eef7a05fb4c53a5b21e03a/pkg/core/pipeline/target/main.go#L157))where all scm relies on [this](https://github.com/updatecli/updatecli/blob/d752c2dbb681613b45eef7a05fb4c53a5b21e03a/pkg/plugins/utils/gitgeneric/main.go#L266)

**Comment by argoyle** [comment]
> > Could it be "as simple as" just updating the local repository after the commit has been made using the API? I cloned the repository and started looking into it but it takes a while to understand the flows.
> 
> I thought It was already done before each target we run a git checkout ([link](https://github.com/updatecli/updatecli/blob/d752c2dbb681613b45eef7a05fb4c53a5b21e03a/pkg/core/pipeline/target/main.go#L157))where all scm relies on [this](https://github.com/updatecli/updatecli/blob/d752c2dbb681613b45eef7a05fb4c53a5b21e03a/pkg/plugins/utils/gitgeneric/main.go#L266)

Hmmm, true. Are the targets running in parallell if no dependencies are specified or something like that? Could it be some kind of race condition perhaps? Saw a couple of errors in an older PR where the wrong commit-SHA was expected or something like that.

**Comment by olblak** [comment]
> Are the targets running in parallell

No we don't, only cloning git repository is done in parallel because it's less risky.
Doing it for sources/condition/target would introduce a level of complexity that I prefer to delay as much as possible 😄

> Could it be some kind of race condition perhaps
Maybe, when we create the commit from [here](https://github.com/updatecli/updatecli/blob/d752c2dbb681613b45eef7a05fb4c53a5b21e03a/pkg/plugins/scms/github/scm.go#L129) we use the wrong base commit hash

**Comment by olblak** [comment]
Maybe @MattiasAng has an idea about where to investigate

**Comment by MattiasAng** [comment]
I think it's something going on with the base commit hash yes. Unfortunately I don't have any time to investigate this currently but I would;

- Set a breakpoint in the [queryHeadOid](https://github.com/updatecli/updatecli/blob/main/pkg/plugins/scms/github/main.go#L497) and check if you're getting the latest commit hash after each individual commit. Could potentially be that the Github API is responding back with an older commit for some reason. If it is responding back with the wrong commit hash then I would test introducing a sleep to see if we might be requesting the API too quickly after the commit.

As said in the original pull request our use-case is extremely simple with just a single SCM and target.

**Comment by MattiasAng** [comment]
@olblak @argoyle 

Okay so I got a bit curious so took 15 minutes to see if I saw anything obvious. I noticed that changes introduced in the first commit wasn't available in the checked out branch on subsequent targets.

I was able to get these changes by changing the [pull options during checkout](https://github.com/updatecli/updatecli/blob/eb6c1f95d529e74235ac7ab4b12dbdd8b581313f/pkg/plugins/utils/gitgeneric/main.go#L343-L346) to:

````go
pullOptions := git.PullOptions{
	RemoteName:    DefaultRemoteReferenceName,
	ReferenceName: plumbing.NewBranchReferenceName(newBranch),
	Force:         true,
	Progress:      &b,
}
````

Could you see if you can build with these changes and test on your end as well?

**Comment by argoyle** [comment]
I tried locally now but get those errors now:
```
ERROR: something went wrong in target "agent_image" : "reference not found"
```

**Comment by MattiasAng** [comment]
Right, it comes down to locally cloned Git directory still being available after a run which means [branch creation](https://github.com/updatecli/updatecli/blob/eb6c1f95d529e74235ac7ab4b12dbdd8b581313f/pkg/plugins/utils/gitgeneric/main.go#L299) is not hit for the first target and fails since the branch is not available in remote repository.

Best fix I can think of is checking if the remote branch exists before trying to pull changes.

```go
if r, _ := repository.Remote(newBranch); r != nil {
	b := bytes.Buffer{}
	pullOptions := git.PullOptions{
		RemoteName:    DefaultRemoteReferenceName,
		ReferenceName: plumbing.NewBranchReferenceName(newBranch),
		Force:         true,
		Progress:      &b,
	}

	if !isAuthEmpty(&auth) {
		pullOptions.Auth = &auth
	}

	err = worktree.Pull(&pullOptions)
	if b.String() != "" {
		logrus.Debugln(b.String())
	}
	b.Reset()
	if err != nil &&
		err != git.ErrNonFastForwardUpdate &&
		err != git.NoErrAlreadyUpToDate {
		logrus.Debugln(err)
		return err
	}
}
```

**Comment by argoyle** [comment]
Still having the problem that changes in one target is not visible (and subsequently reverted in the commit from the next target).
First commit: 
<img width="759" alt="image" src="https://github.com/updatecli/updatecli/assets/328334/330a99f1-b24f-4145-be2d-bbd531482711">
Second commit:
<img width="785" alt="image" src="https://github.com/updatecli/updatecli/assets/328334/b742c8f7-d8bb-46b5-9ea2-8fa304fb7bff">

**Comment by olblak** [comment]
@argoyle I finally had the time to look into this, and I think I found the solution. Could you have another look? 
Thanks @MattiasAng for the useful hints

**Comment by argoyle** [comment]
🎉 Now we're talking. Looks good to me now. Thanks!
